README.md
# ring-jdbc-session

A Clojure library to implement an HTTP session store for
[Ring](https://github.com/ring-clojure/ring) using a JDBC backing store.

Tested with Clojure versions 1.3, 1.4 and 1.5-RC3 with the following databases:

* H2 in-memory database
* MySQL
* PostgreSQL

## Usage

### Leiningen dependency

Coordinates: `[ring-jdbc-session "0.1.0"]`

On Clojars: https://clojars.org/ring-jdbc-session

### Pre-requisite

1. First of all, you need a connection-pooling `javax.sql.DataSource` instance
for use with this library. A `DataSource` is like a connection factory. See
[Clj-DBCP](https://github.com/kumarshantanu/clj-dbcp) (example below uses this) and
[c3p0](http://clojure.github.com/java.jdbc/doc/clojure/java/jdbc/ConnectionPooling.html)
on how to create a `DataSource`. Let's assume the `DataSource` is bound to a var `ds`.

    ```clojure
    (def ds (clj-dbcp.core/make-datasource :mysql {:host "localhost" :database "abc"
                                                   :user "dbuser" :password "s3cr3t"}))
    ```

2. Create the session table if not already created:

    ```clojure
    (ring-jdbc-session.core/create-session-table ds)
    ```

    Or simply create a table using the following
    [DDL](http://en.wikipedia.org/wiki/Data_definition_language):

    ```sql
    CREATE TABLE ring_session (session_key VARCHAR(100) UNIQUE NOT NULL,
                               session_val VARCHAR(1024),
                               session_ts TIMESTAMP NOT NULL)
    ```

### Create and use the session store

#### Use with Ring

```clojure
(let [jdbc-ss (ring-jdbc-session.core/make-session-store ds)]
  (ring-jdbc-session.core/start-cleaner jdbc-ss) ; starting cleaner is optional
  (ring.middleware.session/wrap-session handler {:store jdbc-ss}))
```

#### Use with Compojure

```clojure
(let [jdbc-ss (ring-jdbc-session.core/make-session-store ds)]
  (ring-jdbc-session.core/start-cleaner jdbc-ss) ; starting cleaner is optional
  (compojure.handler/site handler {:session {:store jdbc-ss}}))
```

## Documentation

See file `doc/intro.md` in this repo.

## Getting in touch

On [Ring discussion group](https://groups.google.com/forum/?fromgroups=#!forum/ring-clojure)

On Twitter: [@kumarshantanu](https://twitter.com/kumarshantanu)

## License

Copyright © 2013 Shantanu Kumar

Distributed under the Eclipse Public License, the same as Clojure.


CHANGES.md
# Changes and TODO

* [TODO] Delta calculation/serialization
* [TODO] Text compression

## 2013-Jan-26 / 0.1.0

* Implementation of `ring.middleware.session.store/SessionStore`
* Background stoppable cleaner implementation


test/ring_jdbc_session/core_test.clj
(ns ring-jdbc-session.core-test
  (:require [clojure.pprint :as pp]
            [clj-dbcp.core  :as dbcp]
            [ring-jdbc-session.core :as rjss])
  (:use clojure.test)
  (:import (java.sql ResultSet)))


(def ds nil)


(def opts {})


(defmacro with-stmt
  [bindings & body]
  `(rjss/with-connection (rjss/make-conn-factory ds)
     (rjss/do-prepared ~bindings
                       ~@body)))


(defmacro with-session-table
  [& body]
  `(do
     ;; setup
     (rjss/create-session-table ds opts)
     (try ~@body
       (finally
         ;; teardown
         (rjss/drop-session-table ds opts)))))


(defn print-db-table [table]
  (with-stmt [stmt (str "SELECT * FROM " table)]
    (with-open [^ResultSet rs (.executeQuery stmt)]
      ;;(is (not (.next r)))
      (pp/print-table (resultset-seq rs)))))


(defn count-rows [table]
  (with-stmt [stmt (str "SELECT COUNT(*) AS cnt FROM " table)]
    (with-open [^ResultSet rs (.executeQuery stmt)]
      (.next rs)
      (.getLong rs "cnt"))))


(defn sleep-seconds
  [n]
  (print (format "Sleeping for %d seconds..." n))
  (flush)
  (rjss/sleep (* 1000 n))
  (println "woke up."))


(deftest table-management-works
  (testing "Create session table"
           (rjss/create-session-table ds opts))
  (testing "Read from created session table"
           (with-stmt [stmt "SELECT * FROM ring_session"]
             (.executeQuery stmt)))
  (testing "Drop session table"
           (rjss/drop-session-table ds opts))
  (testing "Read from dropped session table should throw SQLException"
           (is (thrown? java.sql.SQLException
                        (with-stmt [stmt "SELECT * FROM ring_session"]
                          (.executeQuery stmt))))))


(deftest session-store-works
  (with-session-table
    (let [store (rjss/make-session-store ds (merge {:expire-secs 1} opts))]
      (testing "SQL COUNT operation"
               (with-stmt [stmt "SELECT COUNT(*) AS cnt FROM ring_session"]
                 (with-open [^ResultSet rs (.executeQuery stmt)]
                   (is (.next rs))
                   (is (= 0 (.getLong rs "cnt")) "count=0 before tests"))))
      (testing "Write (create) new session key/val"
               (let [struct {:foo "foo" :bar [2 :baz]}]
                 (is (string? (rjss/write store nil "init-value")))
                 (is (= "foo" (rjss/write store "foo" "foo-value")))
                 (is (= "baz" (rjss/write store "baz" struct)))
                 (is (= "foo-value" (rjss/read store "foo")) "create")
                 (is (= struct      (rjss/read store "baz")) "data structure")))
      (testing "Read/write (update) old session key/val"
               (rjss/write store "bar" "bar-value")
               (rjss/write store "bar" "bar-bar")
               (is (= "bar-bar" (rjss/read store "bar")) "update"))
      (testing "Read non-existing session val"
               (is (= nil (rjss/read store "quux")) "non-existing key"))
      (testing "Delete session key/val"
               (rjss/delete store "foo")
               (is (= nil (rjss/read store "foo")) "deleted key"))
      (testing "Read expired session val"
               (sleep-seconds 2)
               (is (= nil (rjss/read store "bar")) "expired key is not retrievable")
               (with-stmt [stmt "SELECT 1 AS present FROM ring_session WHERE session_key=?"]
                 (.setString stmt 1 "bar")
                 (with-open [^ResultSet rs (.executeQuery stmt)]
                   (is (.next rs) "expired key is not deleted"))))
      (testing "Clean stale keys"
               ;; (println "Session table before cleanup") (print-db-table "ring_session")
               (is (= 3 (count-rows "ring_session")) "count=3 before cleanup")
               (rjss/clean store)
               ;; (println "Session table after cleanup") (print-db-table "ring_session")
               (is (= 0 (count-rows "ring_session")) "count=2 after cleanup")))))


(deftest session-cleaning-works
  (with-session-table
    (let [store (rjss/make-session-store ds (merge {:expire-secs 1} opts))]
      (testing "Cleaner"
               (is (= "foo" (rjss/write store "foo" "foo-value")))
               (is (= "bar" (rjss/write store "bar" {:foo 'bar})))
               (is (= 2 (count-rows "ring_session")) "count=2 before cleanup")
               (let [cleaner (rjss/start-cleaner store {:interval-secs 2})]
                 (sleep-seconds 3)
                 (is (= 0 (count-rows "ring_session")) "count=0 after cleanup")
                 (rjss/stop cleaner)
                 (is (true? (.stopped? cleaner)) "Cleaner should be stopped"))
               (is (= "foo" (rjss/write store "foo" "foo-value")))
               (is (= "bar" (rjss/write store "bar" {:foo 'bar})))
               (sleep-seconds 3)
               (is (= 2 (count-rows "ring_session")) "count=2 before cleanup")))))


(deftest non-expiring-session-store-works
  (with-session-table
    (let [store (rjss/make-session-store ds (merge {:expire-secs 0} opts))]
      (testing "Non-expiring keys"
               (is (= "foo" (rjss/write store "foo" "foo-value")))
               (is (= "bar" (rjss/write store "bar" {:foo 'bar})))
               (sleep-seconds 2)
               (is (= 2 (count-rows "ring_session")) "count=2 after sleep")
               (is (= "foo-value" (rjss/read store "foo")))
               (is (= {:foo 'bar} (rjss/read store "bar"))))
      (testing "Cleaner for non-expiring store"
               (let [cleaner (rjss/start-cleaner store {:interval-secs 2})]
                 (sleep-seconds 3)
                 (is (= 2 (count-rows "ring_session")) "count=2 after sleep")
                 (is (= "foo-value" (rjss/read store "foo")))
                 (is (= {:foo 'bar} (rjss/read store "bar")))
                 (rjss/stop cleaner)
                 (is (true? (.stopped? cleaner)) "Cleaner should be stopped"))))))


(defn test-ns-hook []
  (doseq [each (read-string (slurp "test-config/config.clj"))]
    (with-redefs [opts (first each)
                  ds   (apply dbcp/make-datasource (rest each))]
      (println "Running tests using config:" each)
      (table-management-works)
      (session-store-works)
      (session-cleaning-works)
      (non-expiring-session-store-works))))


project.clj
(defproject ring-jdbc-session "0.1.0"
  :description "JDBC based HTTP session store for Ring (Clojure)"
  :url "https://github.com/kumarshantanu/ring-jdbc-session"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :warn-on-reflection true
  :profiles {:dev {:dependencies [[ring/ring-core      "1.1.8"]
                                  [clj-dbcp            "0.8.0"]
                                  [com.h2database/h2   "1.3.170"]
                                  [mysql/mysql-connector-java "5.1.6"]
                                  [postgresql/postgresql      "9.1-901.jdbc4"]]}
             :1.3 {:dependencies [[org.clojure/clojure "1.3.0"]]}
             :1.4 {:dependencies [[org.clojure/clojure "1.4.0"]]}
             :1.5 {:dependencies [[org.clojure/clojure "1.5.0-RC3"]]}}
  :aliases {"all" ["with-profile" "1.3,dev:1.4,dev:1.5,dev"]
            "dev" ["with-profile" "1.4,dev"]})


src/ring_jdbc_session/core.clj
(ns ring-jdbc-session.core
  "JDBC based session store"
  (:refer-clojure :exclude [read])
  (:require [ring.middleware.session.store :as rss])
  (:import java.util.UUID
           (java.sql  Connection PreparedStatement ResultSet Timestamp)
           (javax.sql DataSource)))


(defn new-session-key [prefix]
  (str prefix ":" (str (UUID/randomUUID))))


(defn make-conn-factory [ds-or-cf]
  (cond
    (fn? ds-or-cf) ds-or-cf
    (instance? DataSource ds-or-cf) #(.getConnection ^DataSource ds-or-cf)
    :otherwise (throw (IllegalArgumentException.
                        (str "Expected javax.sql.DataSource or function, found: "
                             (class ds-or-cf) " " ds-or-cf)))))


(def ^{:dynamic true :tag Connection} *connection* nil)


(defmacro with-connection [conn-factory & body]
  (let [conn (gensym)]
    `(with-open [~(with-meta conn {:tag Connection}) (~conn-factory)]
       (binding [*connection* ~conn]
         ~@body))))


(defmacro do-prepared [bindings & body]
  (let [pairs (->> (partition 2 bindings)
                (mapcat (fn [[stmt-var sql-template]]
                          [`~(with-meta stmt-var {:tag PreparedStatement})
                           `(.prepareStatement *connection* ~sql-template)])))]
    `(with-open ~(vec pairs)
       ~@body)))


(defn ^Timestamp ts-now
  ([]
    (ts-now 0))
  ([delta]
    (Timestamp. (+ (System/currentTimeMillis) delta))))


(defn make-ts-fn
  [expire-millis]
  (if (pos? expire-millis)
    #(ts-now (- expire-millis))
    (constantly (Timestamp. 0))))


(defn make-retriever [table-name key-column val-column ts-column expire-millis]
  (let [sql-select (format "SELECT %s FROM %s WHERE %s=? AND %s>=?"
                           val-column table-name key-column ts-column)
        sql-update (format "UPDATE %s SET %s=? WHERE %s=?"
                           table-name ts-column key-column)
        ts-fn      (make-ts-fn expire-millis)]
    (fn [session-key] {:pre [(string? session-key)]}
      (do-prepared [select-stmt sql-select
                    update-stmt sql-update]
                   (doto select-stmt
                     (.setString    1 session-key)
                     (.setTimestamp 2 (ts-fn)))
                   (with-open [^ResultSet rs (.executeQuery select-stmt)]
                     (when (.next rs)
                       (let [r (.getObject rs 1)]
                         ;; update last-accessed timestamp
                         (doto update-stmt
                           (.setTimestamp 1 (ts-now))
                           (.setString    2 session-key)
                           (.executeUpdate))
                         r)))))))


(defn make-persister [table-name key-column val-column ts-column]
  (let [sql-update (format "UPDATE %s SET %s=?, %s=? WHERE %s=?"
                           table-name val-column ts-column key-column)
        sql-insert (format "INSERT INTO %s (%s, %s, %s) VALUES (?, ?, ?)"
                           table-name key-column val-column ts-column)]
    (fn [session-key session-val] {:pre [(string? session-key)]}
      (do-prepared [update-stmt sql-update
                    insert-stmt sql-insert]
                   (doto update-stmt ; update successful?
                     (.setObject    1 session-val)
                     (.setTimestamp 2 (ts-now))
                     (.setString    3 session-key))
                   (when (zero? (.executeUpdate update-stmt))
                     ;; insert
                     (doto insert-stmt
                       (.setString    1 session-key)
                       (.setObject    2 session-val)
                       (.setTimestamp 3 (ts-now))
                       (.executeUpdate))))
      nil)))


(defn make-remover [table-name key-column]
  (let [sql (format "DELETE FROM %s WHERE %s=?" table-name key-column)]
    (fn [session-key] {:pre [(string? session-key)]}
      (do-prepared [stmt sql]
                   (doto stmt
                     (.setString 1 session-key)
                     (.executeUpdate)))
      nil)))


(defn make-cleaner [table-name ts-column expire-millis]
  (let [sql   (format "DELETE FROM %s WHERE %s < ?" table-name ts-column)
        ts-fn (make-ts-fn expire-millis)]
    (fn []
      (do-prepared [stmt sql]
                   (doto stmt
                     (.setTimestamp 1 (ts-fn))
                     (.executeUpdate)))
      nil)))


(defprotocol SessionStoreCleanup
  "Cleanup API for session store"
  (clean-session [_] "Clean up the expired key-val pairs"))


(deftype JdbcStore [conn-factory retriever persister remover cleaner
                    serializer deserializer prefix]
  rss/SessionStore
  (read-session [_ session-key]
    (when session-key  ; read from the DB only when session-key is not nil
      (with-connection conn-factory
        (when-let [val-payload (retriever session-key)]
          (deserializer val-payload)))))
  (write-session [_ session-key data]
    (with-connection conn-factory
      (let [session-key (or session-key (new-session-key prefix))
            val-payload (serializer data)]
        (persister session-key val-payload)
        session-key)))
  (delete-session [_ session-key]
    (with-connection conn-factory
      (remover session-key)))
  SessionStoreCleanup
  (clean-session [_]
    (with-connection conn-factory
      (cleaner))))


;; ========== session store ==========


(defn make-session-store
  "Create a JDBC-backed session storage engine. Argument `conn-factory` can be
  either a javax.sql.DataSource object or a no-arg function that returns a
  java.sql.Connection object."
  ([conn-factory]
    (make-session-store conn-factory {}))
  ([conn-factory {:keys [table-name key-column val-column ts-column
                         serializer deserializer prefix expire-secs]
                  :or {table-name   "ring_session"
                       key-column   "session_key"
                       val-column   "session_val"
                       ts-column    "session_ts"
                       serializer   pr-str
                       deserializer read-string
                       prefix       "session"
                       expire-secs  1800}
                  :as opts}] {:pre [(or (instance? DataSource conn-factory)
                                        (fn? conn-factory))
                                    (string? table-name)
                                    (string? key-column)
                                    (string? val-column)
                                    (string? ts-column)
                                    (fn? serializer)
                                    (fn? deserializer)
                                    (string? prefix)
                                    (number? expire-secs)]}
    (let [retriever (make-retriever table-name key-column val-column ts-column
                                    (* 1000 expire-secs))
          persister (make-persister table-name key-column val-column ts-column)
          remover   (make-remover   table-name key-column)
          cleaner   (make-cleaner   table-name ts-column (* 1000 expire-secs))]
      (JdbcStore. (make-conn-factory conn-factory)
                  retriever persister remover cleaner
                  serializer deserializer prefix ))))


(defn read [session-store session-key] {:pre [(instance? JdbcStore session-store)
                                              (string? session-key)]}
  (.read-session ^JdbcStore session-store session-key))


(defn write [session-store session-key session-val] {:pre [(or (string? session-key)
                                                               (nil? session-key))
                                                           (instance? JdbcStore session-store)]}
  (.write-session ^JdbcStore session-store session-key session-val))


(defn delete [session-store session-key] {:pre [(instance? JdbcStore session-store)
                                                (string? session-key)]}
  (.delete-session ^JdbcStore session-store session-key))


(defn clean [session-store] {:pre [(instance? JdbcStore session-store)]}
  (.clean-session ^JdbcStore session-store))


;; ========== starting/stopping the cleaner ==========


(defprotocol Stoppable
  "Something that can be stopped"
  (stopped? [_] "Return true if stopped, false otherwise")
  (stop     [_] "Stop (idempotent)"))


(defn sleep [millis]
  (let [timeout (+ millis (System/currentTimeMillis))]
    (while (< (System/currentTimeMillis) timeout)
      (try (Thread/sleep (- timeout (System/currentTimeMillis)))
        (catch InterruptedException _
          (.interrupt ^Thread (Thread/currentThread)))))))


(defn start-cleaner
  ([^JdbcStore session-store]
    (start-cleaner session-store {}))
  ([^JdbcStore session-store {:keys [interval-secs]
                              :or {interval-secs 60}
                              :as opts}]
    (let [state-atom  (atom :running)
          store-agent (agent session-store)
          interval-ms (* 1000 interval-secs)
          runner-fn   (fn runner [_]
                        (try
                          (when @state-atom
                            (try
                              (.clean-session session-store)
                              (catch Exception e
                                (.printStackTrace e)))
                            (sleep interval-ms)
                            (send-off store-agent runner))
                          (catch Throwable t
                            (println "Exiting, error occured:" (.getMessage t))
                            (.printStackTrace t))))]
      (send-off store-agent runner-fn)
      (reify Stoppable
        (stopped? [_] (not @state-atom))
        (stop     [_] (swap! state-atom (constantly false)))))))


(defn stop-cleaner [session-cleaner] {:pre [(satisfies? Stoppable session-cleaner)]}
  (.stop session-cleaner))


;; ========== table create/drop stuff ==========


(defn create-session-table
  ([conn-factory]
    (create-session-table conn-factory {}))
  ([conn-factory {:keys [table-name key-column val-column ts-column
                         key-column-type val-column-type ts-column-type]
                  :or {table-name  "ring_session"
                       key-column  "session_key"
                       val-column  "session_val"
                       ts-column   "session_ts"
                       key-column-type "VARCHAR(100) UNIQUE NOT NULL"
                       val-column-type "VARCHAR(1024)"
                       ts-column-type  "TIMESTAMP NOT NULL"}
                  :as opts}]
    (with-connection (make-conn-factory conn-factory)
      (do-prepared [stmt (str "CREATE TABLE " table-name " ("
                              key-column " " key-column-type ", "
                              val-column " " val-column-type ", "
                              ts-column  " " ts-column-type ")")]
                   (.executeUpdate stmt)))))


(defn drop-session-table
  ([conn-factory]
    (drop-session-table conn-factory {}))
  ([conn-factory {:keys [table-name]
                  :or {table-name "ring_session"}
                  :as opts}]
    (with-connection (make-conn-factory conn-factory)
      (do-prepared [stmt (str "DROP TABLE " table-name)]
                   (.executeUpdate stmt)))))
